scripts/main.py

```diff
@@ -0,0 +1,232 @@
+        i = random.choice(range(model.nq))
+        d[i] = random.random() * vel_dir
+        q[i] = q_init
+        x_guess[:, i] = np.linspace(q_init, q_fin, N_guess)
+
+    d /= np.linalg.norm(d)
+    controller.setGuess(x_guess, u_guess)
+
+    # Solve the OCP
+    x_star, u_star, N = controller.solveVBOC(q, d, N_guess, n=N_increment, repeat=50)
+    if x_star is None:
+        return None, None
+    if test_flag:
+        return x_star[0], x_star
+    # Add a final node
+    x_star = np.vstack([x_star, x_star[-1]])
+    # Save the initial state as valid data
+    valid_data = np.vstack([valid_data, x_star[0]])
+
+    # Generate unviable sample in the cost direction
+    x_tilde = np.full((N + 1, model.nx), None)
+    x_tilde[0] = np.copy(x_star[0])
+    x_tilde[0, model.nq:] -= model.eps * d
+
+    x_limit = True if model.checkVelocityBounds(x_tilde[0, model.nq:]) else False
+
+    # Iterate along the trajectory to verify the viability of the solution
+    for j in range(1, N):
+        if x_limit:
+            if model.checkStateBounds(x_star[j]):
+                x_limit = True
+            else:
+
+                if model.checkPositionBounds(x_star[j - 1, :model.nq]):
+                    break
+
+                gamma = np.linalg.norm(x_star[j, model.nq:])
+                d = - x_star[j, model.nq:]
+                d /= np.linalg.norm(d)
+                controller.resetHorizon(N - j)
+                controller.setGuess(x_star[j:N], u_star[j:N])
+                x_new, u_new, _ = controller.solveVBOC(x_star[j, :model.nq], d, N - j, n=N_increment, repeat=5)
+                if x_new is not None:
+                    x0 = controller.ocp_solver.get(0, 'x')
+                    gamma_new = np.linalg.norm(x0[model.nq:])
+                    if gamma_new > gamma + controller.tol:
+                        # Update the optimal trajectory
+                        x_star[j:N], u_star[j:N] = x_new[:N - j], u_new[:N - j]
+
+                        # Create unviable state
+                        x_tilde[j] = np.copy(x_star[j])
+                        x_tilde[j, model.nq:] += model.eps * x_tilde[j, model.nq:] / gamma_new
+
+                        # Check if the unviable state is on bound
+                        x_limit = True if model.checkVelocityBounds(x_tilde[j, model.nq:]) else False
+
+                    else:
+                        x_limit = False
+                        x_tilde[j] = np.copy(x_star[j])
+                        x_tilde[j, model.nq:] -= model.eps * d
+                else:
+                    for k in range(j, N):
+                        if model.checkVelocityBounds(x_star[k, model.nq:]):
+                            valid_data = np.vstack([valid_data, x_star[k]])
+                    break
+        else:
+            x_tilde[j] = simulator.simulate(x_tilde[j - 1], u_star[j - 1])
+            x_limit = True if model.checkStateBounds(x_tilde[j]) else False
+        if model.insideStateConstraints(x_star[j]):
```

Comment:
It is not clear why there are two methods that seem to do the same thing: `checkStateBounds` and `insideStateConstraints`

Reply:
Moreover checking the state constraintson x_star should not be necessary

Reply:
The check at line 105 is taken from the old code, but it seems to introduce a bug, since it excludes the points on the state bounds. I fix it by removing the check. I also remove al the `inside` methods in the `AbstactModel` class.